src/main/java/org/springframework/integration/kafka/core/DefaultConnectionFactory.java

```diff
@@ -172,12 +103,17 @@ public void setFetchMetadataTimeout(int fetchMetadataTimeout) {
 	 */
 	@Override
 	public BrokerAddress getLeader(Partition partition) {
-		try {
-			lock.readLock().lock();
```

Comment:
???
````java
 *       // Must release read lock before acquiring write lock
 *       rwl.readLock().unlock();
 *       rwl.writeLock().lock();
 *       try {
 *         // Recheck state because another thread might have
 *         // acquired write lock and changed state before we did.
 *         if (!cacheValid) {
 *           data = ...
 *           cacheValid = true;
 *         }
 *         // Downgrade by acquiring read lock before releasing write lock
 *         rwl.readLock().lock();
 *       } finally {
 *         rwl.writeLock().unlock(); // Unlock write, still hold read
 *       }
````

I just mean that you have already `getMetadataCache()` instead of that `metadataCacheHolder.get()`. OK, nevermind this case. 

But `readLock` concern is still here...